debian/opus.8
.TH OPUS 8 "11 February 2009" "4.1.0"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
OPUS \- command line interface to the Online Placement university System
.SH SYNOPSIS
.B opus
.SH DESCRIPTION
This manual page documents briefly the
.B OPUS
command line options.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBOPUS\fP is a web application to manage placement and work based learning. It has a command line tool which this manpage documents. The command line tool is usually called by the cron framework for OPUS but can be useful in its own right.
.SH OPTIONS
The syntax for opus commands is
.PP
opus <function> "[parameters]"
.PP
where parameters takes the form used for encoding cgi-scripts. Most functions take no parameters.
.TP
.B start
opens OPUS to non superusers.
.TP
.B stop
prevents access to OPUS for all non superusers.
.TP
.B user_count
show information on the number of users in different categories.
.TP
.B company_count
show the number of companies stored in the database.
.TP
.B vacancy_count
show the number of vacancies storied in the database.
.TP
.B expire_vacancies
finds vacancies past their closing date and closes them, notifying contacts.
.TP
.B phone_home_install
determines if OPUS is permitted to send an email to the University of Ulster on installation, and if so performs this action.
.TP
.B phone_home_periodic
determines if OPUS is permitted to send a regular email to the University of Ulster with item counts, and if it is due, and if so performs this action.
.TP
.B update_perl_config
produced a perl version of the database configuration, which is necessary for timeline generation.
.TP
.B change_password
allows an OPUS user's password to be changed. Parameters are passed in the form "username=foo&password=bar", noting that encapsulation by the quotes is necessary to prevent the ampersand to be misinterpreted by the shell.
.TP
.B auto_add_students
takes a file with one registration number per line and tries, using web services, to add these students into OPUS.
.TP
.B check_missing_error_prompts
used by OPUS developers or translators to check for error prompts used in the code that are not defined or translated.
.TP
.B check_missing_lang_prompts
used by OPUS developers or translators to check for language prompts used in the code that are not defined or translated.

More documentation on OPUS can be found at http://foss.ulster.ac.uk/projects/opus or by installing the opus-doc package on Debian.
.SH AUTHOR
OPUS is produced by the University of Ulster, under a development team led by Colin Turner <c.turner@ulster.ac.uk>.
.PP
This manual page was written by Colin Turner <c.turner@ulster.ac.uk>,
for the Debian project (but may be used by others).


ChangeLog
2008-04-04  colin

	* INSTALL: forgot to mention that an update will be required
	* html/controllers/admin/index_directories.php,
	  include/model/Vacancy.class.php: fix for bug which caused
	  vacancies to jump company if one edited a vacancy for company a
	  when company b was selected
	* include/opus.conf.php: updating version to 4.0.1
	* html/controllers/admin/index_directories.php,
	  include/model/Student.class.php,
	  include/model/StudentImport.class.php,
	  include/model/WebServices.php: students were imported with
	  incorrect usernames, and fixed a problem with note display

2008-04-08  colin

	* etc/apache2.conf: some fixes / improvements to default apache
	  configuration
	* ChangeLog: Adding changelog (4.0.0 to 4.0.1)

2008-04-13  colin

	* configs/lang_en.conf: fix missing prompt for add staff member

2008-04-15  colin

	* configs/lang_en.conf,
	  html/controllers/application/index_application.php,
	  html/controllers/application/index_information.php,
	  html/controllers/application/index_user.php: Improvements to API
	  users, allowing easier interactive use for debugging and testing
	* templates/application/user/decoded_data.tpl,
	  templates/application/user/form_kill_session.tpl: Improvements to
	  API users, allowing easier interactive use for debugging and
	  testing

2008-04-16  colin

	* configs/lang_en.conf,
	  html/controllers/student/index_placement.php,
	  html/controllers/student/index_student.php,
	  include/model/Placement.class.php: students can now view and edit
	  their placement records, should deal with bug #341
	* html/controllers/student/index_student.php: students can now view
	  and edit their placement records, should deal with bug #341
	* configs/lang_en.conf,
	  html/controllers/student/index_placement.php,
	  templates/student/placement/placement_home.tpl: Show most recent
	  placement on student home page
	* html/css/print.css: adding a rudimentary print stylesheet
	* templates/main.tpl: using new print style sheet
	* html/css/print.css: stop tables from being truncated on gecko
	* templates/list.tpl: explicitly specify table headers
	* html/css/print.css: allow table headers to be handled cleverly
	  when printing (if browser supports)

2008-04-17  colin

	* include/model/Placement.class.php, templates/manage.tpl: lookups
	  that are readonly now don't display a drop down

2008-04-18  colin

	* include/model/Student.class.php: Stop username corruption
	  (sometimes become blank).
	* include/dto/DTO_Company.class.php: if a company has no associated
	  activities (impossible in any recent OPUS install) it didn't show
	  in the list (bug #341).
	* include/dto/DTO_Company.class.php: undoing previous fix, it will
	  cause more confusion than it will solve

2008-04-19  colin

	* debian/opus.postinst: remove line to try and install pear log, it
	  is itself Debian packaged and I'll move the dependency to uuwaf
	* debian/control: Ensure that uuwaf 1.0.1 or higher is installed
	  since that now carries the dependency for PEAR Log

2008-04-21  colin

	* debian/opus.postinst: streamlining OPUS packaging

2008-04-22  colin

	* ChangeLog, Makefile, debian/changelog, debian/control,
	  debian/docs: Last changes for 4.0.1
	* html/controllers/admin/index_directories.php: the button to add
	  an administrator got accidentally removed just before 4.0.0
	  release, it is now reinstated (as usual, after you search for an
	  admin).
	* include/opus.conf.php: updating version number

2008-04-24  colin

	* debian/docs: Debian packaging obviously does not require the
	  UPGRADE and INSTALL files, and in fact they produce lintian
	  errors, so removing them from the package.
	* debian/conffiles: Debian automatically treats the contents of etc
	  as conffiles, so duplicating them causes a lintian error
	* INSTALL, UPGRADE, debian/opus.dirs, debian/opus.postinst,
	  include/opus.conf.php: Ensure strict compliance with the
	  Filesystem Hierarchy Standard, by moving photos and resources to
	  /var/lib/opus
	* Makefile, debian/changelog, debian/control, debian/docs: various
	  packaging improvements

2008-04-25  colin

	* debian/copyright: Remove xinha from copyright message
	* html/controllers/admin/index_directories.php: fixed a problem
	  which could cause a fatal error when editing an admin sometimes

2008-04-26  colin

	* include/model/Placement.class.php: removed some inert code and
	  commented file

2008-04-28  colin

	* include/model/Log_Viewer.class.php,
	  include/model/StudentImport.class.php,
	  templates/admin/information/log_viewer.tpl: log viewer should now
	  support compressed (log rotated) log files
	* etc/opus.logrotate.d: including a log rotate configuration now
	  compressed reading is possible

2008-04-29  colin

	* debian/README.Debian, debian/opus.postinst, debian/rules: added
	  initial readme for debian users and also configuration for
	  logrotation, and the possibility to install apache1
	  configurations.

2008-04-30  colin

	* debian/README.Debian, debian/opus.postinst, debian/opus.postrm,
	  debian/opus.templates: more improvements to packaging, a
	  local_en.conf will now also be installed.
	* etc/local_en.conf: initial local_en.conf
	* include/model/Admin.class.php: improved comments
	* Makefile: updated for new debian packaging

2008-05-01  colin

	* templates/manage.tpl: remove overzealous checking for textareas
	  to stop a slowdown problem visible on slower connections with
	  firefox (bug #371)
	* configs/lang_en.conf, debian/changelog,
	  html/controllers/student/index_placement.php,
	  html/controllers/supervisor/index_home.php, html/css/default.css,
	  include/model/Student.class.php,
	  templates/student/placement/placement_home.tpl,
	  templates/supervisor/home/home.tpl: updated version, adding
	  photos of students for supervisors, and academic tutors for
	  students, plus some improved documentation
	* html/controllers/student/index_placement.php,
	  templates/student/placement/placement_home.tpl: small error in

cron/opus.php
#!/usr/bin/php -q
<?php

// php scripts, very irritatingly, use the directory they were called from,
// not the one the script is in. Try to rectify this. We use __FILE__ since
// (at least in my testing), it seems to follow symlinks and still show the
// "real" path.
if(preg_match("%^(.*)/%", __FILE__, $matches)) $directory = $matches[1]; // Unix
// Windows regexp requires four slashes, since PHP requires two for each
// backslash required for the regexp!
if(preg_match("/^(.*)\\\\/", __FILE__, $matches)) $directory = $matches[1]; // Windows

if(!chdir($directory)) die("OPUS: could not change to script directory");

// This include is relative
require_once("../include/opus.conf.php");
// Now we use config to set the others
set_include_path($config['php']['include_dir']);

main();

/**
 * This is the main function that controls OPUS
 *
 * @uses WA.class.php
 * @uses $config
 * @uses WA::request()
 *
 */

function main() 
{
  global $config;
  global $config_sensitive;

  require_once("opus.conf.php");
  require_once("WA.class.php");
  require_once("UUWAF.class.php");

  // Running in batch mode
  $config['waf']['unattended'] = true;

  if($config['opus']['benchmarking'])
  {
    require_once("model/Benchmark.class.php");
    $benchmark = new Benchmark;
  }

  // Initialise the Web Application Framework
  $waf =& UUWAF::get_instance($config['waf']);
  $waf->log("cron job started");

  $waf->assign_by_ref("benchmark", $benchmark);
  $waf->assign_by_ref("config", $config);


  $waf->register_data_connection('default', $config_sensitive['opus']['database']['dsn'], $config_sensitive['opus']['database']['username'], $config_sensitive['opus']['database']['password']);

  $waf->create_log_file("cron");
  $waf->set_default_log("cron");
  $waf->set_log_ident("cron");

  $function = $_SERVER['argv'][1];

  // other parameters passed in regular var=value&var2=value2 format
  $parameters = $_SERVER['argv'][2];
  $parameters = explode("&", $parameters);
  $parameter_count = 0;
  $key = array();
  $value = array();
  foreach($parameters as $keyvalue)
  {
    $parts = explode("=", $keyvalue);
    $key[$parameter_count] = $parts[0];
    $value[$parameter_count] = $parts[1];
    $parameter_count++;
  }
  $decoded_parameters = array_combine($key, $value);
  $waf->assign("parameters", $decoded_parameters);

  if(function_exists($function))
  {
    $waf->log("invoking $function", PEAR_LOG_DEBUG);
    $function($waf, $decoded_parameters);
  }
  else
  {
    help($waf, $parameters);
  }
  echo "\n";
}

function help(&$waf, $parameters)
{
  $waf->display("cron.tpl", "cron:user_count", "cron/help.tpl");
}

function dev_help(&$waf, $parameters)
{
  $waf->display("cron.tpl", "cron:user_count", "cron/dev_help.tpl");
}

function check_online_users(&$waf, $parameters)
{
  require_once("model/User.class.php");
  User::check_online_users();
  User::drop_online_user_count_file();
}

function user_count(&$waf, $parameters)
{
  require_once("model/User.class.php");
  $roots       = User::count("where user_type='root'");
  $admins      = User::count("where user_type='admin'");
  $contacts    = User::count("where user_type='company'");
  $staff       = User::count("where user_type='staff'");
  $supervisors = User::count("where user_type='supervisors'");
  $students    = User::count("where user_type='student'");

  $waf->assign("roots", $roots);
  $waf->assign("admins", $admins);
  $waf->assign("contacts", $contacts);
  $waf->assign("staff", $staff);
  $waf->assign("supervisors", $supervisors);
  $waf->assign("students", $students);

  $waf->display("cron.tpl", "cron:user_count", "cron/user_count.tpl");
}

function company_count(&$waf, $parameters)
{
  require_once("model/Company.class.php");
  $companies = Company::count();

  $waf->assign("companies", $companies);

  $waf->display("cron.tpl", "cron:user_count", "cron/company_count.tpl");
}

function vacancy_count(&$waf, $parameters)
{
  require_once("model/Vacancy.class.php");
  $vacancies = Vacancy::count();

  $waf->assign("vacancies", $vacancies);

  $waf->display("cron.tpl", "cron:user_count", "cron/vacancy_count.tpl");
}

function phone_home_install(&$waf, $parameters)
{
  require_once("model/PhoneHome.class.php");

  PhoneHome::send_install();
}

function phone_home_periodic(&$waf, $parameters)
{
  require_once("model/PhoneHome.class.php");

  PhoneHome::send_periodic();
}

/*
function create_admin(&$waf, $parameters)
{
  require_once("model/Admin.class.php");
  $admin = new Admin;

  $fields['username'] = $parameters['username'];
  $fields['password'] = md5($parameters['password']);
  $fields['salutation'] = "Dr";
  $fields['firstname'] = "Demo";
  $fields['lastname'] = "User";
  print_r($fields);
  Admin::insert($fields);
}
*/

function start()
{
  require_once("model/Service.class.php");

  Service::start();
}

function stop()
{
  require_once("model/Service.class.php");

  Service::stop();
}

function update_timelines($waf, $parameters)
{
  // Better ensure the configuration is up-to-date
  update_perl_config();

  require_once("model/Timeline.class.php");
  Timeline::update_all_years();
}

function update_timelines_for_year($waf, $parameters)
{
  // Better ensure the configuration is up-to-date
  update_perl_config();

  require_once("model/Timeline.class.php");
  // Do a very comprehensive check
  Timeline::update_year($parameters['year'], true);
}

function expire_vacancies($waf, $parameters)
{
  require_once("model/Vacancy.class.php");
  Vacancy::close_expired_vacancies();
}

/**
* write the database access credentials to a suitable perl file
*/
function update_perl_config()
{
  $waf =& UUWAF::get_instance();
  global $config_sensitive;

  $filename = "config.pl";
  $fp = @fopen($filename, "w");
  if($fp == false)
  {
    $message = "unable to write perl configuration file";
    echo $message;
    $waf->log($message);
    return;
  }
  fwrite($fp, "# automatically generated file, make changes to the php configuration\n");
  fwrite($fp, "# and run 'opus update_perl_config'\n\n");
  fwrite($fp, "package opus;\n\n");
  fwrite($fp, "\$db_dsn='dbi:" . $config_sensitive['opus']['database']['dsn'] . "';\n");
  fwrite($fp, "\$db_username='" . $config_sensitive['opus']['database']['username'] . "';\n");
  fwrite($fp, "\$db_password='" . $config_sensitive['opus']['database']['password'] . "';\n");
  fwrite($fp, "\nreturn 1;\n# end of file\n");
  fclose($fp);
}

function get_academic_year()
{
  global $config;
  $yearstart = $config['opus']['yearstart'];
  if(empty($yearstart)) $yearstart="0930";

  if(empty($year)){
    if(date("md") < $yearstart) $year = date("Y") - 1;
    else $year = date("Y");
  }
  return($year);
}

function change_password($waf, $parameters)
{
  if(empty($parameters['username']) || (empty($parameters['password'])))
  {
    help($waf);
    return;
  }
  if(posix_getuid())
  {
    echo "must be run as root.";
    return;
  }
  
  echo "Changing password for " . $parameters['username'] . " ... ";
  
  $waf->log("change password for user " . $parameters['username']);
  require_once("model/User.class.php");
  $user = User::load_by_username($parameters['username']);
  if($user->id)
  {
    // Successfully loaded a real user
    $user->password = md5($parameters['password']);
    $user->_update();
    echo "success\n";
  }
  else
  {
    $waf->log("password change failed, user could not be loaded.");
    echo "failed\n";
  }
}

function auto_add_students($waf, $parameters)
{
  if(empty($parameters['filename']))
  {
    help($waf);
    return;
  }
  if(posix_getuid())
  {
    echo "must be run as root.";
    return;
  }

  require_once("model/StudentImport.class.php");
  StudentImport::auto_add_by_file($parameters['filename']);
}

function check_missing_error_prompts()
{
  $halt_command = "grep -Ihor \"halt(\\\"error:.*.*\\\")\" ../* | sed -e \"s/halt(\\\"//\" -e \"s/\\\")//\" | sort | uniq";
  // Look for the error codes called on halt, from the top of the source tree
  $fp = popen($halt_command, 'r');
  while(!feof($fp))
  {
    $line = trim(fgets($fp));
    if(empty($line)) continue; // Last line is blank
    // grep for this now!
    

    $missing_halt_command = "grep -IL \"" . $line . "\" ../configs/lang_*";
    $missing_files = array();
    $fp_missing = popen($missing_halt_command, "r");
    while(!feof($fp_missing))
    {
      $missing_files_line = trim(fgets($fp_missing));
      if(!empty($missing_files_line)) array_push($missing_files, $missing_files_line);
    }
    pclose($fp_missing);
    if(count($missing_files))
    {
      echo "Prompt $line... missing for :\n";
      foreach($missing_files as $missing_files_line)
      {
        echo "  $missing_files_line\n";
      }
    }
    
  }
  pclose($fp);
}

function check_missing_lang_prompts()
{
  $lang_command = "grep -Ihor \\\"[a-z_]*:[a-z_]*:[a-z_]*:[a-z_]*\\\" ../* | sort | uniq";
  // Look for the error codes called on halt, from the top of the source tree
  $fp = popen($lang_command, 'r');
  while(!feof($fp))
  {
    $line = trim(fgets($fp));
    if(empty($line)) continue; // Last line is blank
    // grep for this now!
    

    $missing_lang_command = "grep -IL \"" . $line . "\" ../configs/lang_*";
    $missing_files = array();
    $fp_missing = popen($missing_lang_command, "r");
    while(!feof($fp_missing))
    {
      $missing_files_line = trim(fgets($fp_missing));
      if(!empty($missing_files_line)) array_push($missing_files, $missing_files_line);
    }
    pclose($fp_missing);
    if(count($missing_files))
    {
      echo "Prompt $line... missing for :\n";
      foreach($missing_files as $missing_files_line)
      {
        echo "  $missing_files_line\n";
      }
    }
    
  }
  pclose($fp);
}

?>